artifact.go

```diff
@@ -203,67 +213,423 @@ func (m artifactModule) getBuildTestsResultsHandler(c *gin.Context) {
+	if !ok {
+		return
+	}
+
+	details := []TestResultDetail{}
+	err := m.Database.
+		Where(&TestResultDetail{BuildID: uint(buildID)}).
+		Find(&details).
+		Error
 
-	form, err := c.MultipartForm()
 	if err != nil {
-		ginutil.WriteMultipartFormReadError(c, err, fmt.Sprintf(
-			"Failed reading multipart-form content from request body when uploading new artifact for build with ID %d.",
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result details for build with ID %d from database.",
 			buildID))
 		return
 	}
 
+	c.JSON(http.StatusOK, details)
+}
+
+// getBuildTestResultDetailsHandler godoc
+// @summary Get test result details for specified test
+// @tags artifact
+// @param buildid path int true "Build ID"
+// @param artifactId path int true "Artifact ID"
+// @success 200 {object} []TestResultDetail
+// @router /build/{buildid}/test-result-details/{artifactId} [get]
+func (m artifactModule) getBuildTestResultDetailsHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, buildIDParamName)
+	if !ok {
+		return
+	}
+
+	artifactID, ok := ginutil.ParseParamUint(c, artifactIDParamName)
+	if !ok {
+		return
+	}
+
+	details := []TestResultDetail{}
+	err := m.Database.
+		Where(&TestResultDetail{BuildID: uint(buildID), ArtifactID: artifactID}).
+		Find(&details).
+		Error
+
+	if err != nil {
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result details from test with ID %d for build with ID %d from database.",
+			artifactID, buildID))
+		return
+	}
+
+	c.JSON(http.StatusOK, details)
+}
+
+// getBuildTestResultsSummaryHandler godoc
+// @summary Get test result summary of all tests for specified build
+// @tags artifact
+// @param buildid path int true "Build ID"
+// @success 200 {object} TestResultListSummary
+// @failure 400 {object} problem.Response "Bad Request"
+// @failure 502 {object} problem.Response "Bad Gateway"
+// @router /build/{buildid}/test-results-summary [get]
+func (m artifactModule) getBuildTestResultsSummaryHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, buildIDParamName)
+	if !ok {
+		return
+	}
+
+	summaries := []TestResultSummary{}
+	err := m.Database.
+		Preload("Artifact").
+		Where(&TestResultSummary{BuildID: uint(buildID)}).
+		Find(&summaries).
+		Error
+
+	if err != nil {
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result summaries for build with ID %d from database.",
+			buildID))
+		return
+	}
+
+	summaryList := TestResultListSummary{
+		BuildID:   buildID,
+		Summaries: summaries}
+
+	for _, v := range summaries {
+		summaryList.Failed += v.Failed
+		summaryList.Passed += v.Passed
+		summaryList.Skipped += v.Skipped
+	}
+
+	summaryList.Total = summaryList.Failed + summaryList.Passed + summaryList.Skipped
+
+	c.JSON(http.StatusOK, summaryList)
+}
+
+func parseMultipartFormData(c *gin.Context, buildID uint) ([]file, bool) {
+	files := make([]file, 0)
```

Comment:
Suggest moving this declaration to just above the `for` loop where it's used. And you don't need to `make` up a new slice, as nil slices and empty slices are treated equally.

```go
append([]int{}, 1) // => []int{ 1 }
append(make([]int), 2) // => []int{ 2 }
append([]int(nil), 3) // => []int{ 3 }

len([]int{}) // => 0
len(make([]int)) // => 0
len([]int(nil)) // => 0

// the following are semantically identical:
var files []file
var files []file = nil
var files []file = []file(nil)
var files        = []file(nil)
```

```suggestion
	var files []file
```